Mathlib/Tactic/NormNum/NormNumI.lean

```diff
@@ -0,0 +1,228 @@
+
+theorem IsComplex.neg : ∀ {z : ℂ} {a b : ℝ}, IsComplex z a b → IsComplex (-z) (-a) (-b)
+  | _, _, _, ⟨rfl⟩ => ⟨rfl⟩
+
+theorem IsComplex.conj : ∀ {z : ℂ} {a b : ℝ}, IsComplex z a b → IsComplex (conj z) a (-b)
+  | _, _, _, ⟨rfl⟩ => ⟨rfl⟩
+
+theorem IsComplex.ofNat (n : ℕ) [n.AtLeastTwo] :
+    IsComplex (OfNat.ofNat (α := ℂ) n) (OfNat.ofNat n) 0 := ⟨rfl⟩
+
+theorem IsComplex.scientific (m exp : ℕ) (x : Bool) :
+    IsComplex (OfScientific.ofScientific m x exp : ℂ) (OfScientific.ofScientific m x exp : ℝ) 0 :=
+  ⟨rfl⟩
+
+theorem eq_eq {z : ℂ} {a b a' b' : ℝ} (pf : IsComplex z a b) (pf_a : a = a') (pf_b : b = b') :
+  IsComplex z a' b' := by simp_all
+
+theorem eq_of_eq_of_eq_of_eq {z w : ℂ} {az bz aw bw : ℝ}
+    (hz : IsComplex z az bz) (hw : IsComplex w aw bw)
+    (ha : az = aw) (hb : bz = bw) : z = w := by
+  simp [hz.out, hw.out, ha, hb]
+
+theorem ne_of_re_ne {z w : ℂ} {az bz aw bw : ℝ} (hz : IsComplex z az bz) (hw : IsComplex w aw bw)
+    (ha : az ≠ aw) : z ≠ w := by
+  simp [hz.out, hw.out, ha]
+
+theorem ne_of_im_ne {z w : ℂ} {az bz aw bw : ℝ} (hz : IsComplex z az bz) (hw : IsComplex w aw bw)
+    (hb : bz ≠ bw) : z ≠ w := by
+  simp [hz.out, hw.out, hb]
+
+theorem IsComplex.re_eq {z : ℂ} {a b : ℝ} (hz : IsComplex z a b) : Complex.re z = a := by
+  simp [hz.out]
+
+theorem IsComplex.im_eq {z : ℂ} {a b : ℝ} (hz : IsComplex z a b) : Complex.im z = b := by
+  simp [hz.out]
+
+theorem IsComplex.of_pow_negSucc {z w : ℂ} {a b : ℝ} {n : ℕ} {k k' : ℤ} (h : k = Int.negSucc n)
+    (hk : NormNum.IsInt k' k)
+    (hz : IsComplex (w ^ (n + 1))⁻¹ a b) (hz' : z = w ^ (k' : ℤ)) : IsComplex z a b := by
+  obtain rfl : k' = k := by simpa using hk.out
+  constructor
+  simpa [h, hz'] using hz.out
+
+theorem IsComplex.of_pow_ofNat {z w : ℂ} {k k' : ℤ} {n : ℕ} {a b : ℝ} (hz1 : z = w ^ k)
+    (hw : IsComplex (w ^ n) a b) (hk' : k' = n) (hkk' : NormNum.IsInt k k') :
+    IsComplex z a b := by
+  obtain rfl : k = k' := by simpa using hkk'.out
+  constructor
+  simpa [hz1, hk'] using hw.out
+
+/-- Parsing all the basic calculation in complex. -/
+partial def parse (z : Q(ℂ)) : MetaM (Σ a b : Q(ℝ), Q(IsComplex $z $a $b)) := do
+  match z with
+  /- parse an addition: `z₁ + z₂` -/
+  | ~q($z₁ + $z₂) =>
+    let ⟨_a₁, _b₁, pf₁⟩ ← parse z₁
+    let ⟨_a₂, _b₂, pf₂⟩ ← parse z₂
+    pure ⟨_, _, q(.add $pf₁ $pf₂)⟩
+  /- parse a multiplication: `z₁ * z₂` -/
+  | ~q($z₁ * $z₂) =>
+    let ⟨_a₁, _b₁, pf₁⟩ ← parse z₁
+    let ⟨_a₂, _b₂, pf₂⟩ ← parse z₂
+    pure ⟨_, _, q(.mul $pf₁ $pf₂)⟩
+  /- parse an inversion: `z⁻¹` -/
+  | ~q($z⁻¹) =>
+    let ⟨_x, _y, pf⟩ ← parse z
+    pure ⟨_, _, q(.inv $pf)⟩
+  /- parse `z₁/z₂` -/
+  | ~q($z₁ / $z₂) => do
+    let ⟨_a, _b, pf⟩ ← parse q($z₁ * $z₂⁻¹)
+    return ⟨_, _, q($pf)⟩
+  /- parse `-z` -/
+  | ~q(-$w) => do
+    let ⟨_a, _b, pf⟩ ← parse w
+    return ⟨_, _, q(.neg $pf)⟩
+  /- parse a subtraction `z₁ - z₂` -/
+  | ~q($z₁ - $z₂) => parse q($z₁ + -$z₂)
```

Comment:
This isn't type safe, and is abusing the defeq. Better to define a separate `IsComplex.sub`.